flang/lib/Lower/PFTBuilder.cpp

```diff
@@ -369,6 +410,89 @@ class PFTBuilder {
     evaluationListStack.pop_back();
   }
 
+  /// Rewrite IfConstructs containing a GotoStmt to eliminate an unstructured
+  /// branch and a trivial basic block.  The pre-branch-analysis code:
+  ///
+  ///       <<IfConstruct>>
+  ///         1 If[Then]Stmt: if(cond) goto L
+  ///         2 GotoStmt: goto L
+  ///         3 EndIfStmt
+  ///       <<End IfConstruct>>
+  ///       4 AssignmentStmt: ...
+  ///       5 AssignmentStmt: L ...
+  ///
+  /// becomes:
+  ///
+  ///       <<IfConstruct>>
+  ///         1 If[Then]Stmt [negate]: if(cond) goto L
+  ///         4 AssignmentStmt: ...
+  ///         3 EndIfStmt
+  ///       <<End IfConstruct>>
+  ///       5 AssignmentStmt: L ...
+  ///
+  /// The If[Then]Stmt condition is implicitly negated.  It is not modified
+  /// in the PFT.  It must be negated when generating FIR.  The GotoStmt is
+  /// deleted.
+  ///
+  /// The transformation is only valid for forward branch targets at the same
+  /// construct nesting level as the IfConstruct.  The result must not violate
+  /// construct nesting requirements or contain an EntryStmt.  The result
+  /// is subject to normal un/structured code classification analysis.  The
+  /// result is allowed to violate the F18 Clause 11.1.2.1 prohibition on
+  /// transfer of control into the interior of a construct block, as that does
+  /// not compromise correct code generation.  When two transformation
+  /// candidates overlap, at least one must be disallowed.  In such cases,
+  /// the current heuristic favors simple code generation, which happens to
+  /// favor later candidates over earlier candidates.  That choice is probably
+  /// not significant, but could be changed.
+  ///
+  void rewriteIfGotos() {
+    using T = struct {
+      lower::pft::EvaluationList::iterator ifConstructIt;
+      parser::Label ifTargetLabel;
+    };
+    llvm::SmallVector<T, 8> ifExpansionStack;
+    auto &evaluationList = *evaluationListStack.back();
+    for (auto it = evaluationList.begin(), end = evaluationList.end();
+         it != end; ++it) {
+      auto &eval = *it;
+      if (eval.isA<parser::EntryStmt>()) {
+        ifExpansionStack.clear();
+        continue;
+      }
+      auto firstStmt = [](lower::pft::Evaluation *e) {
+        return e->isConstruct() ? &*e->evaluationList->begin() : e;
+      };
+      auto &targetEval = *firstStmt(&eval);
+      if (targetEval.label) {
+        while (!ifExpansionStack.empty() &&
+               ifExpansionStack.back().ifTargetLabel == *targetEval.label) {
+          auto ifConstructIt = ifExpansionStack.back().ifConstructIt;
+          auto successorIt = std::next(ifConstructIt);
+          if (successorIt != it) {
+            auto &ifBodyList = *ifConstructIt->evaluationList;
+            ifBodyList.erase(std::next(ifBodyList.begin())); // GotoStmt
```

Comment:
I wonder if you might want an `assert` here to make sure this is erasing a Goto. In case the algorithm is changed (poorly) and this goes haywire.